services/src/main/java/org/keycloak/services/resources/admin/AdminConsole.java

```diff
@@ -232,28 +233,36 @@ public Response whoAmI(final @Context HttpHeaders headers) {
     private void addRealmAccess(RealmModel realm, UserModel user, Map<String, Set<String>> realmAdminAccess) {
         RealmManager realmManager = new RealmManager(session);
         ClientModel realmAdminApp = realm.getClientByClientId(realmManager.getRealmAdminClientId(realm));
-        getRealmAdminAccess(realmAdminApp, user, realmAdminAccess);
+        getRealmAdminAccess(realm, realmAdminApp, user, realmAdminAccess);
     }
 
     private void addMasterRealmAccess(UserModel user, Map<String, Set<String>> realmAdminAccess) {
         List<RealmModel> realms = session.realms().getRealms();
         for (RealmModel realm : realms) {
             ClientModel realmAdminApp = realm.getMasterAdminClient();
-            getRealmAdminAccess(realmAdminApp, user, realmAdminAccess);
+            getRealmAdminAccess(realm, realmAdminApp, user, realmAdminAccess);
         }
     }
 
-    private void getRealmAdminAccess(ClientModel client, UserModel user, Map<String, Set<String>> realmAdminAccess) {
-        Set<String> roles = client.getRolesStream().filter(user::hasRole)
-                .map(RoleModel::getName).collect(Collectors.toSet());
 
-        if (!roles.isEmpty()) {
-            if (!realmAdminAccess.containsKey(realm.getName())) {
-                realmAdminAccess.put(realm.getName(), roles);
-            } else {
-                realmAdminAccess.get(realm.getName()).addAll(roles);
-            }
+    private static <T> HashSet<T> union(Set<T> set1, Set<T> set2) {
+        if (set1 == null && set2 == null) {
+            return null;
+        }
+        HashSet<T> res = (set1 instanceof HashSet) ? (HashSet <T>) set1 : new HashSet<>(set1);
```

Comment:
Isn't there a NPE waiting to happen here if set1 is null? IIRC the HasSet constructor raises an NPE if the collection specified in the constructor is null.

Reply:
Good catch, indeed!

Reply:
Good catch, indeed!